src/EditorFeatures/CSharp/MisplacedUsingDirectives/MisplacedUsingDirectivesCodeFixProvider.cs

```diff
@@ -0,0 +1,328 @@
+
+            // Read the preffered placement option and verify if it can be applied to this code file.
+            // There are cases where we will not be able to fix the diagnostic and the user will need to resolve
+            // it manually.
+            var (placement, preferPreservation) = DeterminePlacement(compilationUnit, options);
+            if (preferPreservation)
+            {
+                return;
+            }
+
+            foreach (var diagnostic in context.Diagnostics)
+            {
+                context.RegisterCodeFix(
+                    new MoveMisplacedUsingsCodeAction(token => GetTransformedDocumentAsync(document, compilationUnit, options, placement, token)),
+                    diagnostic);
+            }
+        }
+
+        private static async Task<Document> GetTransformedDocumentAsync(
+            Document document,
+            CompilationUnitSyntax compilationUnit,
+            OptionSet options,
+            AddImportPlacement placement,
+            CancellationToken cancellationToken)
+        {
+            var syntaxFactsService = document.GetLanguageService<ISyntaxFactsService>();
+
+            // Expand usings so that they can be properly simplified after they are relocated.
+            var compilationUnitWithExpandedUsings = (CompilationUnitSyntax)await ExpandUsingDirectivesAsync(document, compilationUnit, cancellationToken).ConfigureAwait(false);
+
+            // Remove the file header from the compilation unit so that we do not lose it when making changes to usings.
+            var (compilationUnitWithoutHeader, fileHeader) = RemoveFileHeader(compilationUnitWithExpandedUsings, syntaxFactsService);
+
+            var newCompilationUnit = placement == AddImportPlacement.InsideNamespace
+                ? MoveUsingsInsideNamespace(compilationUnitWithoutHeader)
+                : MoveUsingsOutsideNamespace(compilationUnitWithoutHeader);
+
+            // Re-attach the header now that using have been moved and LeadingTrivia is no longer being altered.
+            var newCompilationUnitWithHeader = AddFileHeader(newCompilationUnit, fileHeader);
+            var newDocument = document.WithSyntaxRoot(newCompilationUnitWithHeader);
+
+            // Simplify usings now that they have been moved and are in the proper context.
+            return await Simplifier.ReduceAsync(newDocument, Simplifier.Annotation, options, cancellationToken).ConfigureAwait(false);
+        }
+
+        private static async Task<SyntaxNode> ExpandUsingDirectivesAsync(Document document, CompilationUnitSyntax containerNode, CancellationToken cancellationToken)
+        {
+            // Get all using directives so they can be expanded at one time.
+            var namespaceDeclarations = GetMembers(containerNode).OfType<NamespaceDeclarationSyntax>();
+            var allUsingDirectives = namespaceDeclarations.SelectMany(GetUsingDirectives).Concat(containerNode.Usings);
```

Comment:
Can we just use `containerNode.DescendantNodes` to get all `UsingDirectiveSyntax`?